---
title: "Human trophectoderm RNA-seq:<br>excluding aneuploidy by eSNPKaryotyping analysis"
author: "Dr David Iles; OmicsLtd.com"
date: "08/03/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Workflow for eSNPkaryotyping of human trophectoderm RNA-seq data.

### Scripts written to generate vcf files from HISAT2 alignments and create tables of dbSNP151 depth of coverage.

```{bash alignment_processing_to_vcf, echo=TRUE, eval=FALSE}
# remove PCR and optical duplicates

  cd ~/path/to/hisat2/*_mps.bam

  for file in *bam; do 

  of=$(echo ${file} | sed 's/_mps.bam//')
  echo ${of} 

  samtools flagstat ${file} > ${of}.stat

  java -Xmx8g -jar /path/to/picard/picard.jar MarkDuplicates \
    REMOVE_DUPLICATES=true \
    OPTICAL_DUPLICATE_PIXEL_DISTANCE=2500 \
    INPUT=${file} \
    OUTPUT=${of}.noDups.bam \
    METRICS_FILE=${of}_Picard.metrics

  done

# scripts to add read group information, split alignments at N cigar strings and call variants

  for file in *bam; do

  of=$(echo ${file} | sed 's/_noDups.bam//')

  java -Xmx8g -jar ~/path/to/picard/picard.jar AddOrReplaceReadGroups \
    I=${file} \
    O=${of}_noDupsRG.bam \
    RGID=1 \
    RGLB=RPM \
    RGPL=ILLUMINA \
    RGPU=HCKGLBBXX \
    RGSM=trophectoderm_RNAseq

  samtools index ${of}_noDupsRG.bam ${of}_noDupsRG.bam.bai

  ~/path/to/gatk-4.1.0.0/gatk --java-options "-Xmx8G" SplitNCigarReads \
  -R /path/to/human/genome_indexes/hg38/hg38.fa \
  -I ${of}_noDupsRG.bam \
  -O ${of}_noDupsRG.split.bam

  ~/path/to/gatk-4.1.0.0/gatk --java-options "-Xmx8G" HaplotypeCaller \
  -R /path/to/human/genome_indexes/hg38/hg38.fa \
  -I ${of}_noDupsRG.split.bam \
  -O ${of}.vcf.gz

  done

# process SNP data for individual chromosomes against UCSC hg38.chrom.sizes

  cd ~/path/to/where/hg38_commonSNPs151.gtf/file/is

  awk 'BEGIN {IFS = OFS = "\t"} {print $1, $4, $5}' 
    hg38_commonSNPs151_19Feb19.gtf > hg38_commonSNPs151_19Feb19.tab
  
  awk 'BEGIN {IFS = OFS = "\t"} $2 == $3 {print $1, $2, $3}' \
    hg38_commonSNPs151_19Feb19.tab > hg38_commonSNPs151_SNP-only_27Feb19.tab

  sort -k1,1 -k2,2n \
  hg38_commonSNPs151_SNP-only_27Feb19.tab > hg38_commonSNPs151_SNP-only_27Feb19s.tab


# and then trim to cover only those chromsomes in the primary assembly 
# (omitting Alt-, Un- and random etc)

  intersectBed \
  -a hg38_commonSNPs151_SNP-only_27Feb19s.tab \
  -b hg38.chrom1-22XY.bed > hg38_commonSNPs151_SNP-only_1-22XY_27Feb19.tab

# finally, calculate depth of chromosomal coverage for each ${of}_noDupsRG.split.bam 
# within intervals defined in hg38.chrom1-22XY.bed,
# and select SNPs with coverage depth >= 20

  for file in *bam; do

  of=$(echo ${file} | sed 's/_noDupsRG.split.bam//')
  echo "processing "${file}

  samtools depth -b hg38.chrom1-22XY.bed ${file} > ${of}_reads-per-position.txt
  echo "filtering "${file}

  awk 'BEGIN {FS = "\t"} $3 >= 20 {print $0}' \
    ${of}_reads-per-position.txt > ${of}_reads-per-position2.txt

  done

```

### eSNP karyotyping in R.

#### Routines to process variant calls and generate genome plots of allelic ratios and hetero/homozygosity.

```{r karyotyping, echo=TRUE, eval=FALSE}
  R
  library(zoo)
  library(gplots)

# extract SNP calls from vcf files

  setwd('~/path/to/the/HaplotypeCaller/vcf/files')
  files <- list.files(pattern = '*vcf.gz$')

  vcf.list <- lapply(files, function(x){
    dat = read.table(x, as.is=T)[,c(1,2,10)]
    colnames(dat) <- c('chr', 'position', 'info')
    dat$chr <- gsub('chr', '', dat$chr)
    dat$chr <- gsub('X', '23', dat$chr)
    dat$chr <- gsub('Y', '24', dat$chr)
    dat$chr <- as.numeric(dat$chr)
    dat <- dat[!is.na(dat$chr),]

    infoRegex = "^([01])\\/([01]):(\\d+)\\,(\\d+):(\\d+):\\d+:\\d+\\,\\d+\\,\\d+$"

    dat$Karyotype <- 10*abs(as.numeric(gsub(infoRegex, "\\1", dat$info)) - 
                              as.numeric(gsub(infoRegex, "\\2", dat$info)))
    dat <- dat[!is.na(dat$Karyotype),]
    dat$AD1 <- as.numeric(gsub(infoRegex, "\\3", dat$info))
    dat$AD2 <- as.numeric(gsub(infoRegex, "\\4", dat$info))
    dat$DP <- as.numeric(gsub(infoRegex, "\\5", dat$info))
    dat <- dat[,c(1,2,5:7,4)]
    dat <- dat[order(dat$chr, dat$position),]
    dat
    })

  names(vcf.list) <- gsub('.vcf.gz', '', files)

# scripts to call major/minor allele frequencies
# function to calculate major/minor allele frequencies

  source('~/path/to/Sort_major_minor.R')
  source('~/path/to/calcMM.R')

# process vcf data in vcf.list and calculate allele frequencies
  
  MajMin.list <- list()
  MajMin.list <- lapply(vcf.list, MajorMinorCalc)
  names(MajMin.list) <- names(vcf.list)
```
The following two routines were adapted from eSNPKaryotyping scripts to detect regions exhibiting regions of potential duplication and/or deletion along human autosomes.

#### Plotting allelic ratios along chromosomes.

```{r plotting_allelic_ratios, echo=TRUE, eval=FALSE}
  source('~/path/to/PlotAutosomalAllelicRatioFDR.R')
  
  for (i in 1:length(MajMin.list)){
    tab <- MajMin.list[[i]]
    PlotAutosomalAllelicRatioFDR(tab, window = 151, Tlim = 3)
    dev.copy2eps(file = paste0(names(MajMin.list[i]), '_ARautosomePlot_', 
                             format(Sys.Date(), "%d%b%y"), '.eps'))
    dev.off()
  }

# simulating trisomy 18 in implanted sample 2
    tab <- MajMin.list[[2]]
    tab$MajorMinor <- ifelse(tab$chr == 18, tab$MajorMinor + 0.7, tab$MajorMinor)
    tab$MajorMinor <- ifelse(tab$MajorMinor > 4, 4, tab$MajorMinor)
    PlotAutosomalAllelicRatioFDR(tab, window = 151, Ylim = 3)
    dev.copy2eps(file='BRP_IMPL_2_simulated_trisomy18_genomePlot_13Mar19.eps')
    dev.off()

```
#### Plotting zygosity along chromosomes.

```{r plotting_zygosity, echo=TRUE, eval=FALSE}
# load common dbSNP151 data
  snps151 <- read.table('~/path/to/hg38_commonSNPs151_SNP-only_1-22XY_27Feb19.tab', 
                        header=F, sep='\t')
  colnames(snps151) <- c('chr', 'start', 'end')
  snps151$chr <- gsub('chr', '', snps151$chr)
  snps151$chr <- gsub('X', '23', snps151$chr)
  snps151$chr <- gsub('Y', '24', snps151$chr)
  snps151$chr <- as.numeric(snps151$chr)
  snps151 <- snps151[!is.na(snps151$chr),]

# plot SNP heterozygosity and coverage from MajMin.list across the genome 
# using the PZSautosomes function
# first load coverage files
  rpp2.files <- list.files(path='~/path/to/*noDupsRGsplit_bam/', pattern='*2.txt')

  source('~/path/to/PZSautosomes.R')

  for(i in 1:length(MajMin.list)) {
    rpp <- read.table(rpp2.files[i], header=F, sep='\t')
    colnames(rpp)=c("chr","start","Depth")
    rpp$chr <- gsub('chr', '', rpp$chr)
    rpp$chr <- gsub('X', '23', rpp$chr)
    rpp$chr <- gsub('Y', '24', rpp$chr)
    rpp$chr <- as.numeric(rpp$chr)
    rpp <- rpp[!is.na(rpp$chr),]
    rpp2snps <- merge(rpp, snps151, by=c('chr', 'start'))
    
    mm <- MajMin.list[[i]]
    
    # merge observed heterozygous SNPs with covered dbSNP151
    # using NAs to classify SNP type
    mm=merge(mm, rpp2snps, by=c('chr', 'start'), all=TRUE)
    
    # detected SNPs not present in dbSNP151
    mm$snpType <- 0
    
    # dbSNP151 SNPs covered but NOT heterozygous
    mm$snpType <- ifelse(is.na(mm$AF1), paste(-1), paste(mm$snpType))
    
    # dbSNP151 SNPs covered AND heterozygous
    mm$snpType <- ifelse(!is.na(mm$AF1) & !is.na(mm$Depth), paste(1), paste(mm$snpType))
    mm <- mm[order(mm$chr, mm$start),]
    
    # set up graphics device 
    x11(width = 9, height = 5)
    PZSautosomes(mm)
    dev.copy2eps(file=paste0(names(MajMin.list)[i], '_singleSNPs_autosomePlot_', 
                             format(Sys.Date(), "%d%b%y"), '.eps'))
    dev.off()
    rm(rpp, mm, rpp2snps, i)
  }
  
```
Et voila! :)

*Dr David Iles*
*daveiles@omicsltd.com*

```{r session_info, echo=TRUE, eval=TRUE}
sessionInfo()
```